pkg/client/typed/scale/scale.go

```diff
@@ -0,0 +1,218 @@
+	}
+	switch res.TypeMeta.APIVersion {
+	case "extensions/v1beta1":
+		return &extScale{scale{c, res}}, nil
+	case "autoscaling/v1":
+		return &asScale{scale{c, res}}, nil
+	default:
+		return nil, fmt.Errorf("unknown APIVersion `%s` for `Scale`", res.TypeMeta.APIVersion)
+	}
+}
+
+func (c *Client) Get(scaleRef extensions.SubresourceReference, namespace string) (ScaleInterface, error) {
+	gv, err := unversioned.ParseGroupVersion(scaleRef.APIVersion)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing group version from scale reference: %v", err)
+	}
+	gvr, _ := meta.KindToResource(gv.WithKind(scaleRef.Kind))
+
+	resource := &unversioned.APIResource{
+		Name:       gvr.Resource,
+		Namespaced: true,
+		Kind:       scaleRef.Kind,
+	}
+	res, err := c.dc.Subresource(resource, scaleRef.Subresource, namespace).Get(scaleRef.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch resource %s(%s): %v", scaleRef.Name, scaleRef.Kind, err)
+	}
+	return c.decode(res)
+}
+
+func (c *Client) Update(scaleRef extensions.SubresourceReference, namespace string, obj ScaleInterface) (ScaleInterface, error) {
+	gv, err := unversioned.ParseGroupVersion(scaleRef.APIVersion)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing group version from scale reference: %v", err)
+	}
+	gvr, _ := meta.KindToResource(gv.WithKind(scaleRef.Kind))
+
+	resource := &unversioned.APIResource{
+		Name:       gvr.Resource,
+		Namespaced: true,
+		Kind:       scaleRef.Kind,
+	}
+	var res *runtime.Unstructured
+	switch t := obj.(type) {
+	case *extScale:
+		res, err = c.dc.Subresource(resource, scaleRef.Subresource, namespace).Update(t.getScale())
+	case *asScale:
+		res, err = c.dc.Subresource(resource, scaleRef.Subresource, namespace).Update(t.getScale())
+	default:
+		return nil, fmt.Errorf("error obtaining existing scale values")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch resource %s(%s): %v", scaleRef.Name, scaleRef.Kind, err)
+	}
+	return c.decode(res)
+}
+
+func field(obj map[string]interface{}, keyChain ...string) (interface{}, error) {
+	res := reflect.ValueOf(obj).Interface()
+	for _, key := range keyChain {
+		resVal, ok := res.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("%v should be of type map[string]interface{} to be able to proceed", res)
+		}
+		res, ok = resVal[key]
+		if !ok {
+			return nil, fmt.Errorf("key(`%s`) doesn't exist in the `Scale` object", key)
+		}
+	}
+	return res, nil
+}
+
+type scale struct {
+	c   *Client
+	obj *runtime.Unstructured
+}
+
+func (s *scale) getScale() *runtime.Unstructured {
+	return s.obj
+}
+
+func (s *scale) Replicas() (int, error) {
+	ri, err := field(s.obj.Object, "status", "replicas")
+	if err != nil {
+		return 0, err
+	}
+
+	// Note: JSON only has a number type whose equivalent is float64 in Go. So we type-assert
+	// to float64 here and then cast the value to int before returning.
+	r, ok := ri.(float64)
```

Comment:
> Why does this need to be different than what we do in the rest of the API, where we use integers?

@bgrant0607 As @liggitt said, it is different here because I am decoding the JSON here using Go's json.Unmarshal() as opposed to our regular codec that handles all these details for us.

> I'm concerned that this client's decode is losing existing API typing/defaulting/conversion mechanisms.

@liggitt It is, but what can we do?